tex/chProgramming.tex

```diff
@@ -1192,6 +1208,108 @@ \subsection{Recursion on natural numbers}\label{ssec:recnat}
 notations introduced by the file.
 There is no better way to browse the library than reading these banners.}
 
+\section{Total functions}\label{sec:total}
+When discussing the subtraction function in Section~\ref{ssec:recnat},
+we were confronted with a striking difference between the usual
+mathematical approach to functions and the approach used in \Coq{} and
+the \mcbMC{} library.  In the latter context, {\em all functions are total
+  by construction}.  This is a difficulty when we want the language we
+use in the \mcbMC{} library to stick as close to possible to the
+language of usual mathematics, where some functions are {\em not
+  defined everywhere}.
+
+The functions from Section~\ref{ssec:recnat} where the difference is
+visible are function~\C{predn} and function~\C{subn}.  The former
+returns the predecessor of natural number, and intuitively, it should
+not be defined when this natural number is \C{O}, which has no
+predecessor.  The latter returns the subtraction of a number from
+another one, and intuitively it should not be defined when the first
+argument (from which subtraction is performed) is smaller than the
+amount being subtracted.
+
+Usual mathematical practice, learned by
+everybody from middle school, is that we can consider a bigger type of
+numbers, in which the natural numbers are injected implicitly, and
+where the predecessor of \C{0} and the results of exceptional
+subtractions do exist.  So the predecessor of 3 is a natural number,
+but the predecessor of \C{0} is not, it is a relative integer.  This
+versatility, allowed by mathematical practice, may seem to provide
+the necessary soupleness to be able to write mathematical formulas
+quickly and meaningfully.  It is not the case.  there are mathematical
+functions that are inherently not defined everywhere, in ways that
+cannot be recovered by just adding a new type.
```

Comment:
There are several ring structures in the Math-comp library.  The "unit ring structure" is one where one explicitly decide to have a multiplicative inverse function and a predicate "unit" which describes precisely those elements for which the result of the function is meaningful.  Not all rings need to implement the "unit ring structure", but having this structure independently from the Field structure is useful.